File: backend/middleware.go
package main

import (
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/gin-contrib/cors"
	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
)

func corsMiddleware() gin.HandlerFunc {
	allowedOrigin := os.Getenv("ALLOWED_ORIGIN")

	if allowedOrigin == "" {
		return gin.HandlerFunc(func(c *gin.Context) {
			c.Next()
		})
	}

	return cors.New(cors.Config{
		AllowOrigins: []string{allowedOrigin},
		AllowMethods: []string{"GET", "POST", "PUT", "DELETE"},
		AllowHeaders: []string{"Origin", "Content-Type", "Authorization"},
		MaxAge:       12 * time.Hour,
	})
}

func loggingMiddleware() gin.HandlerFunc {
	return gin.Logger()
}

func recoveryMiddleware() gin.HandlerFunc {
	return gin.Recovery()
}

func authMiddleware(container *Container) gin.HandlerFunc {
	return func(g *gin.Context) {
		authHeader := g.GetHeader("Authorization")
		if authHeader == "" {
			container.Logger.Error("[authMiddleware] No Authorization header")
			g.AbortWithStatusJSON(http.StatusUnauthorized, nil)
			return
		}

		parts := strings.Split(authHeader, " ")
		if len(parts) != 2 || parts[0] != "Bearer" {
			container.Logger.Error("[authMiddleware] Invalid Authorization header format")
			g.AbortWithStatusJSON(http.StatusUnauthorized, nil)
			return
		}

		accessToken := parts[1]

		claims := jwt.MapClaims{}
		jwtSecret := os.Getenv("SUPABASE_JWT_SECRET")
		token, err := jwt.ParseWithClaims(accessToken, claims, func(token *jwt.Token) (interface{}, error) {
			return []byte(jwtSecret), nil
		})

		if err != nil || !token.Valid {
			container.Logger.Error("[authMiddleware] Invalid or expired token")
			g.AbortWithStatusJSON(http.StatusUnauthorized, nil)
			return
		}

		g.Set("user", claims)
		g.Next()
	}
}

File: node-backend/src/middleware.ts
import { FastifyRequest, FastifyReply } from "fastify";
import { JwtPayload } from "@supabase/supabase-js";
import jwt from "jsonwebtoken";
import logger from "./logger.js";

export function authMiddleware(
  request: FastifyRequest,
  reply: FastifyReply,
  done: () => void
) {
  const authHeader = request.headers["authorization"];

  if (!authHeader) {
    logger.error("[authMiddleware] No Authorization header");
    reply.code(401).send();
    return;
  }

  const parts = authHeader.split(" ");
  if (parts.length !== 2 || parts[0] !== "Bearer") {
    logger.error("[authMiddleware] Invalid Authorization header format");
    reply.code(401).send();
    return;
  }

  const accessToken = parts[1];
  const jwtSecret = process.env.SUPABASE_JWT_SECRET as string;
  try {
    request.user = jwt.verify(accessToken, jwtSecret) as JwtPayload;
    request.token = accessToken;
    done();
  } catch (err) {
    logger.error("[authMiddleware] Invalid or expired token");
    reply.code(401).send();
  }
}
